src/scheduler/buckets.c

```diff
@@ -0,0 +1,1249 @@
+dup_node_bucket_array(node_bucket **old, server_info *nsinfo) {
+	node_bucket **new;
+	int i;
+	if (old == NULL)
+		return NULL;
+
+	new = malloc((count_array((void**) old) + 1) * sizeof(node_bucket *));
+	if (new == NULL) {
+		log_err(errno, __func__, MEM_ERR_MSG);
+		return NULL;
+	}
+
+	for (i = 0; old[i] != NULL; i++) {
+		new[i] = dup_node_bucket(old[i], nsinfo);
+		if (new[i] == NULL) {
+			free_node_bucket_array(new);
+			return NULL;
+		}
+	}
+
+	new[i] = NULL;
+
+	return new;
+}
+/* node_bucket destructor */
+void
+free_node_bucket(node_bucket *nb) {
+	if(nb == NULL)
+		return;
+	
+	free_bucket_bitpool(nb->busy);
+	free_bucket_bitpool(nb->busy_later);
+	free_bucket_bitpool(nb->free);
+	
+	free_resource_list(nb->res_spec);
+	
+	pbs_bitmap_free(nb->bkt_nodes);
+
+	free(nb->name);
+	free(nb);
+}
+/* node bucket array destructor */
+void
+free_node_bucket_array(node_bucket **buckets) {
+	int i;
+
+	if (buckets == NULL)
+		return;
+
+	for (i = 0; buckets[i] != NULL; i++)
+		free_node_bucket(buckets[i]);
+
+	free(buckets);
+}
+
+/* node_bucket_count constructor */
+node_bucket_count *
+new_node_bucket_count() {
+	node_bucket_count *nbc;
+
+	nbc = malloc(sizeof(node_bucket_count));
+	if(nbc == NULL) {
+		log_err(errno, __func__, MEM_ERR_MSG);
+		return NULL;
+	}
+	nbc->bkt = NULL;
+	nbc->chunk_count = 0;
+
+	return nbc;
+}
+
+void
+free_node_bucket_count(node_bucket_count *nbc) {
+	if(nbc == NULL)
+		return;
+
+	free(nbc);
+}
+
+void free_node_bucket_count_array(node_bucket_count **nbc_array) {
+	int i;
+	if(nbc_array == NULL)
+		return;
+
+	for(i = 0; nbc_array[i] != NULL; i++)
+		free_node_bucket_count(nbc_array[i]);
+
+	free(nbc_array);
+}
+
+/**
+ * @brief find the index into an array of node_buckets based on resources, queue, and priority
+ * @param[in] buckets - the node_bucket array to search
+ * @param[in] rl - the resource list of the node bucket
+ * @param[in] qinfo - the queue of the node bucket
+ * @param[in] priority - the priority of the node bucket
+ * @return int
+ * @retval index of array if found
+ * @retval -1 if not found or on error
+ */
+int
+find_node_bucket_ind(node_bucket **buckets, schd_resource *rl, queue_info *qinfo, int priority) {
+	int i;
+	if (buckets == NULL || rl == NULL)
+		return -1;
+
+	for (i = 0; buckets[i] != NULL; i++) {
+		if (compare_resource_avail_list(buckets[i]->res_spec, rl)) {
+			if (buckets[i]->queue == qinfo && buckets[i]->priority == priority)
+				return i;
+		}
+	}
+	return -1;
+}
+
+/**
+ * @brief create a name for a node bucket based on resource names, priority, and queue
+ *
+ * @return char *
+ * @retval name of bucket
+ * @retval NULL - error
+ */
+char *create_node_bucket_name(status *policy, node_bucket *nb) {
+	char *name;
+	int len;
+
+	if(policy == NULL || nb == NULL)
+		return NULL;
+
+	name = create_resource_signature(nb->res_spec, policy->resdef_to_check_no_hostvnode, ADD_ALL_BOOL);
+	if (name == NULL)
+		return NULL;
+
+	len = strlen(name);
+
+	if (nb->priority != 0) {
```

Comment:
PBS defines a node priority as a number between -1024 and 1024.  I check if it is not 0 because the scheduler gives a node 0 priority if the priority attribute is unset.  I don't want priority to be in the name in that case.